### PR TITLE
Memory Pair Game

### PR DESCRIPTION
**Pull request by DmytryjK**
### Memory Pair Game

[Demo](https://dmytryjk.github.io/MemoryPairGame/) |
[Code base](https://github.com/DmytryjK/kottans-frontend/tree/main/task_memory_pair_game)

### **Tests were conducted among the students of the courses.**
**Fixed the following bugs:**
- When you click on the "Click to continue" button and hide the popup, there is a shift in the victory message with a picture from the center of the screen. (made a winning message centered in any condition)
- The selection of cards on the field was allowed. (prohibited the selection of cards)
- There was too little time to show two different cards. (added a little more time)
- When you click on the picture again, it closes.
- There was no timer for the current game time. (added current time + best time)
- When choosing 2 identical cards on the field, they visually disappeared, but you could click on them. This added the value of the visually disappeared card to the array for comparison, which led to the closing of the new card on the field. she was compared to the one that had already disappeared.

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by DmytryjK** [comment]
Everything should be okay.

**Comment by DmytryjK** [comment]
I'm still waiting for the review of my work))

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    bestResultsSheet = [];
+let timerInterval;
+
+setTimeout(showAllCards, 400);
+setTimeout(hideAllCards, 2000);
+
+if (localStorage.getItem('bestTime') !== null) {
+  bestTime.innerText = localStorage.getItem('bestTime');
+  bestResultsSheet.push(bestTime.innerText);
+}
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:52]: :fire:

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                }
+              });
+            } else {
+              bestTime.innerText = localStorage.getItem('time');
+            }
+
+            localStorage.setItem('bestTime', bestTime.innerText);
+            clearInterval(timerInterval);
+
+            buttonToContinue.addEventListener('click', () => {
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:98]: Move event listeners callbacks into independent functions, it's hard to read nested functions. Also, remember one thing one function one action/responsibility.

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Awesome design!

File: submissions/DmytryjK/memory-pair-game/js/script.js (hunk 2 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    });
+
+    activeCards = [];
+}
+
+function showAllCards() {
+  listCards.forEach(item => {
+    item.classList.add('active-card');
+    item.style.opacity = "1";
+    item.style.cssText = 'pointer-events: none';
```
[the remark this answers is not in the document]
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js]: fixed

**Comment by DmytryjK** [comment]
I'm not sure about the correctness of the decision taken, but when refactoring the code, I removed the restriction on flipping only 2 cards.
It seems to me that it will be more convenient for the user.

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        });
+
+      } else {
+        bestTime.innerText = localStorage.getItem('time');
+      }
+
+      localStorage.setItem('bestTime', bestTime.innerText);
+      clearInterval(timerInterval);
+
+      buttonToContinue.addEventListener('click', () => {
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:105]: When user has a strict with 5(for example) games in a row he will have 5 event listeners at the end. Remove event listeners after the ending game.

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Good progress!
Remove empty file: variables.scss.
When the user selects a lot of cards he can see more than 2 cards/turn, please fix it
[scrnli_10_1_2022_3-35-24 PM.webm](https://user-images.githubusercontent.com/16196199/193409999-7aa0eefc-e86d-4c41-b677-c1a7589fadd9.webm)

**Comment by DmytryjK** [comment]
> Good progress! Remove empty file: variables.scss. When the user selects a lot of cards he can see more than 2 cards/turn, please fix it [scrnli_10_1_2022_3-35-24 PM.webm](https://user-images.githubusercontent.com/16196199/193409999-7aa0eefc-e86d-4c41-b677-c1a7589fadd9.webm)

Thank's.
A little higher, I wrote that I decided to remove the restriction on opening only 2 cards, it seems to me that such an experience would be better for the user. But I can revert the changes back and prohibit the selection of more than 2 cards.

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+function hideTwoCards(activeCards) {
+  activeCards.forEach(activeCard => {
+    activeCard.classList.add('hide-card');
+  });
+
+  numbersOfCloseCards++;
+}
+
+function closeTwoCards(activeCards) {
```
[the remark this answers is not in the document]
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:127]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        setTimeout(showAllCards, 400);
+        setTimeout(hideAllCards, 2000);
+      });
+
+    }
+    activeCards = [];
+  }
+});
+
+function hideTwoCards(activeCards) {
```
[the remark this answers is not in the document]
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:119]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js (hunk 4 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:105]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  timerInterval = setInterval(timer, 1000);
+}
+
+function createElementsOfCards() {
+  doubleElementsOfCards = [...elementsOfCards, ...elementsOfCards];
+  doubleElementsOfCards.sort(() => Math.random() - 0.5);
+
+  for (let i = 0; i < (doubleElementsOfCards.length); i++) {
+
+    parrentCardBlock.innerHTML += `
```
[the remark this answers is not in the document]
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:156]: fixed

**Comment by DmytryjK** [comment]
Sorry, I found 1 bug with the addition of new cards, it will be necessary to fix the hanging of the event listener.
Unsubscribe, as I correct it, I requested a review early (((

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+  } else if (seconds < 10) {
+    currentTime.innerText = `${minutes} m : 0${seconds} s`;
+  } else if (minutes < 10) {
+    currentTime.innerText = `0${minutes} m : ${seconds} s`;
+  } else {
+    currentTime.innerText = `${minutes} m : ${seconds} s`;
+  }
+}
+
+function resetVariables() {
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:198]: I'm not sure that the current function name describes all provided functionality: move removeEventListeners to another function or rename the current function.

File: submissions/DmytryjK/memory-pair-game/index.html [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+<body>
+    <main class="main">
+        <ul class="main__cards"></ul>
+        <div class="winMessage">
+            <h2 class="winMessage__title">You're a winner!</h2>
+            <img class="winMessage__image" src="images/winMessage.png" alt="">
+            <button class="click-to-continue">click to continue</button>
+        </div>
+    </main>
+    <section class="timer">
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/index.html:20]: On some resolutions, you have a bug with container width: [video](https://share.vidyard.com/watch/93BjC3vaKWEtzDucubBVBy?)

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+      localStorage.setItem('bestTime', bestTime.innerText);
+      clearInterval(timerInterval);
+
+      buttonToContinue.addEventListener('click', clickToContinue, false);
+    }
+    activeCards = [];
+  }
+}
+
+function clickToContinue(buttonToContinue) {
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:116]:
```suggestion
function restartGame(buttonToContinue) {
```
What do you think about that?

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
One more try)

File: submissions/DmytryjK/memory-pair-game/js/script.js (hunk 10 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:116]: I think it makes sense to rename the function. The name you suggested better describes what the function does.

File: submissions/DmytryjK/memory-pair-game/index.html (hunk 9 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/index.html:20]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js (hunk 8 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:198]: fixed

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+  activeCards = [];
+}
+
+function showAllCards() {
+  currentListCards.forEach(card => {
+    card.classList.add('active-card');
+  });
+}
+
+function hideAllCards() {
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:121]: The user still can see more than 2 cards by clicking the rest of the cards. please set restrictions on that)
[scrnli_10_15_2022_11-15-08 AM.webm](https://user-images.githubusercontent.com/16196199/195976727-0e46e68c-fdde-4034-beb9-8ce367ee516c.webm)

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
One last issue)

File: submissions/DmytryjK/memory-pair-game/js/script.js (hunk 11 above)
Comment (DmytryjK) [reply submissions/DmytryjK/memory-pair-game/js/script.js:121]: I know what you mean, but earlier I write about this, I decided not to limit the user to choose only 2 cards on the field, I think this is better for his user experience)
But I can take it all back

File: submissions/DmytryjK/memory-pair-game/js/script.js [hunk 12]
```diff
@@ -75,14 +75,17 @@ function onClickUserActions(e) {
   }
 
   if (activeCards.length >= 2) {
+    
+    currentListCards.forEach(card => {
+      card.classList.add('pointer-events');
```
Comment (A-Ostrovnyy) [inline submissions/DmytryjK/memory-pair-game/js/script.js:80]: Class `pointer-events` does not describe his functionality please avoid this naming in the future.

**Review by A-Ostrovnyy — APPROVED** [review]
Well done!

**Pull request merged**